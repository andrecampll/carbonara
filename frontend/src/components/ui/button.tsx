import { Button as BaseButton, ButtonProps } from '@mui/base/Button'
import clsx from 'clsx'
import * as React from 'react'

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => {
    const { className, ...other } = props
    return (
      <BaseButton
        ref={ref}
        className={clsx(
          'ui-disabled:text-slate-700 ui-disabled:dark:text-slate-200 ui-disabled:bg-slate-200 ui-disabled:dark:bg-slate-700 ui-disabled:cursor-default ui-disabled:shadow-none ui-disabled:dark:shadow-none ui-disabled:hover:bg-slate-200 ui-disabled:hover:dark:bg-slate-700 ui-disabled:border-none hover:bg-secondary-600 active:bg-secondary-700 cursor-pointer rounded-lg border border-solid border-secondary-500 bg-secondary-500 px-4 py-2 font-sans text-sm font-semibold leading-normal text-white transition focus-visible:outline-none active:scale-[0.99] active:shadow-none',
          className,
        )}
        {...other}
      />
    )
  },
)

Button.displayName = 'Button'
