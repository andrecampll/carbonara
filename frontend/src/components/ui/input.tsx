import { Input as _Input } from '@mui/base/Input'
import { forwardRef, InputHTMLAttributes, ReactNode } from 'react'

type InputProps = {
  icon?: ReactNode
  label?: string
  error?: string
} & InputHTMLAttributes<HTMLInputElement>

export const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ id, error, label, ...props }, ref) => {
    return (
      <div className="w-full">
        {label && (
          <label className="text-sm" htmlFor={id}>
            {label}
          </label>
        )}

        <_Input
          slotProps={{
            input: {
              className:
                'w-full text-sm font-sans font-normal leading-5 px-3 py-2 rounded-lg shadow-md shadow-slate-100 focus:shadow-outline-purple focus:shadow-lg border border-solid border-slate-300 hover:border-purple-500 focus:border-purple-500 bg-white text-slate-900 focus-visible:outline-0',
            },
          }}
          ref={ref}
          {...props}
        />

        <div>
          {error && <span className="text-xs text-red-400">{error}</span>}
        </div>
      </div>
    )
  },
)

Input.displayName = 'Input'
