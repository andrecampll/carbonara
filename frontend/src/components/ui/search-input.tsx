import { Autocomplete } from '@mui/material'
import { forwardRef } from 'react'

import { Input, InputProps } from './input'

type Props = InputProps & {
  onChangeValue: (value: string) => void
}

export const SearchInput = forwardRef<HTMLDivElement, Props>(
  ({ label, helperText, onChangeValue }, ref) => {
    const dietaryChoices = [
      {
        value: 'Vegan',
        label: 'Vegan',
      },
      {
        value: 'Vegetarian',
        label: 'Vegetarian',
      },
      {
        value: 'Pescatarian',
        label: 'Pescatarian',
      },
      {
        value: 'MeatEater',
        label: 'Meat Eater',
      },
    ]

    return (
      <div>
        <Autocomplete
          options={dietaryChoices}
          renderInput={(params) => (
            <Input {...params} label={label} helperText={helperText} />
          )}
          onChange={(_, option) => option && onChangeValue(option.value)}
          ref={ref}
          isOptionEqualToValue={(option, value) => option.value === value.value}
        />
      </div>
    )
  },
)

SearchInput.displayName = 'SearchInput'
