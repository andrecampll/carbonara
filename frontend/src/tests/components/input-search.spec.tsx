import { cleanup, fireEvent, render, screen } from '@testing-library/react'
import { afterEach, describe, expect, test, vi } from 'vitest'

import { SearchInput } from '../../components/ui/search-input'

afterEach(cleanup)

describe('SearchInput', () => {
  test('it should be able to render SearchInput component', () => {
    const onChangeValue = () => {}

    const { container } = render(<SearchInput onChangeValue={onChangeValue} />)

    expect(screen.getByRole('combobox')).toBeDefined()
    expect(screen.getByRole('button')).toHaveProperty('title', 'Open')

    expect(container.firstChild).toMatchSnapshot()
  })

  test('it should be able click and expand Search Input', () => {
    const onChangeValue = vi.fn()

    const { container } = render(<SearchInput onChangeValue={onChangeValue} />)

    fireEvent.click(
      screen.getAllByRole('button', {
        name: /open/i,
      })[0],
    )

    const option = screen.getByText(/vegan/i)

    expect(option).toBeDefined()

    fireEvent.click(option)

    expect(container.firstChild).toMatchSnapshot()
  })
})
