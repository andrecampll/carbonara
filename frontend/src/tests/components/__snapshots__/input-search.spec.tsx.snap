// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchInput > it should be able click and expand Search Input 1`] = `
<div
  class="MuiAutocomplete-root Mui-focused MuiAutocomplete-fullWidth MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-5r88sx-MuiAutocomplete-root"
>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1wnwmuc-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-autocomplete="list"
        aria-expanded="false"
        aria-invalid="false"
        autocapitalize="none"
        autocomplete="off"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r2:"
        role="combobox"
        spellcheck="false"
        type="text"
        value="Vegan"
      />
      <div
        class="MuiAutocomplete-endAdornment css-p1olib-MuiAutocomplete-endAdornment"
      >
        <button
          aria-label="Clear"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
          tabindex="-1"
          title="Clear"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="CloseIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="Open"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
          tabindex="-1"
          title="Open"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-ihdtdm"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`SearchInput > it should be able to render SearchInput component 1`] = `
<div
  class="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasPopupIcon css-5r88sx-MuiAutocomplete-root"
>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1wnwmuc-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-autocomplete="list"
        aria-expanded="false"
        aria-invalid="false"
        autocapitalize="none"
        autocomplete="off"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r0:"
        role="combobox"
        spellcheck="false"
        type="text"
        value=""
      />
      <div
        class="MuiAutocomplete-endAdornment css-p1olib-MuiAutocomplete-endAdornment"
      >
        <button
          aria-label="Open"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
          tabindex="-1"
          title="Open"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-ihdtdm"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
